doc: glossary (#46)
awslabs/metrique: metrique/README.md

- [`#[metrics]` macro reference](https://docs.rs/metrique/0.1/metrique/unit_of_work/attr.metrics.html)

Unlike many popular metric frameworks that are based on the concept of your application having a fixed-ish set of counters and gauges, which are periodically updated to a central place, metrique is based on the concept of structured **metric records**. Your application emits a series of metric records - that are essentially structured log entries - to an observability service such as [Amazon CloudWatch], and the observability service allows you to view and alarm on complex aggregations of the metrics.

The log entries being structured means that you can easily use problem-specific aggregations to track down the cause of issues, rather than only observing the symptoms.

[Amazon CloudWatch]: https://docs.aws.amazon.com/AmazonCloudWatch

## Getting Started (Applications)

Most metrics your application records will be "unit of work" metrics. In a classic HTTP server, these are typically tied to the request/response scope.

> You declare a struct the represents the metrics you plan to capture over the course of the request and annotate it with `#[metrics]`. That will make them implement the [`CloseEntry`] trait, which represents an entry that can be "closed" into an [`InflectableEntry`]. The macro will also expose a wrapper `<MetricName>Guard` type (a type alias to [`AppendAndCloseOnDrop`]) that implicitly appends to a sink when the struct is dropped.

Reviewer: ```suggestion
You declare a struct that represents the metrics you plan to capture over the course of the request and annotate it with `#[metrics]`. That makes it possible to write it to a `Sink`. Rather than writing to the sink directly, you typically use `append_on_drop(sink)` to obtain a guard that will automatically write to the sink when dropped.
```

> The best way to emit the entry is by emitting it to a global entry sink, defined by using the [`metrique_writer::sink::global_entry_sink`] macro. That will create a global rendezvous point - you can attach a destination by using [`attach`] or [`attach_to_stream`], and then write to it by using the [`sink`] method (you must attach a destination before calling [`sink`], otherwise you will encounter a panic!).

Reviewer: ```suggestion
The simplest way to emit the entry is by emitting it to a global entry sink, defined by using the [`metrique_writer::sink::global_entry_sink`] macro. That will create a global rendezvous point - you can attach a destination by using [`attach`] or [`attach_to_stream`], and then write to it by using the [`sink`] method (you must attach a destination before calling [`sink`], otherwise you will encounter a panic!).

Typically applications have a single sink. One sink can handle different `#[metric]` structs.
```

## Getting Started (Applications)

Most metrics your application records will be "unit of work" metrics. These are typically tied to the request/response scope. You declare a struct the represents the metrics you plan to capture over the course of the request and annotate it with `#[metrics]`.

The `#[metrics]` macro will make your struct implement the [`CloseEntry`] trait, which represents an entry that
can be "closed" into an [`InflectableEntry`].

> The macro will also expose a wrapper `<MetricName>Guard` type (a type alias to [`AppendAndCloseOnDrop`]) that implicitly appends to a global queue when the struct is dropped. This wrapper in turn exposes other APIs to further tune behavior.

Reviewer: this seems too low level for the getting started? I suspect many people will use library and never know what an `InflectableEntry` is. The key thing to capture in these docs is that instead of having some sort of global counters like you might expect from a metric library, you emit individual entries.

> The best way to emit the [`InflectableEntry`] is by defining a global entry sink using the [`metrique_writer::sink::global_entry_sink`] macro, which defines a global variable that provides a global rendezvous point where metrics can be written into. Then, you need to attach to that global entry sink an actual entry sink that will actually write your metrics somewhere - the example below will write the metrics to an [`tracing_appender::rolling::RollingFileAppender]

Reviewer: the key here is that there is the `global` destination which must be separately wired to a destination. I think if we said something like:

```
Callers can write to the global queue with `::queue()` and wire the queue to a destination with `::attach_to_stream()`.

The queue must be wired to a destination before metrics can be sent to the queue
```

fn initialize_metrics(service_log_dir: PathBuf) -> AttachHandle {

>     // attach a BackgroundQueue EntrySink to our global entry sink,

Reviewer: ```suggestion
    // attach a EMF-formatted rolling file appender to `ServiceMetrics`
```

#[tokio::main]
async fn main() {
    // not strictly needed, but metrique will emit tracing errors

>     // in some condition, and it's best to be able to see them.

Reviewer: ```suggestion
    // when entries are invalid and it's best to be able to see them.
```
